Name: <redacted> <redacted>

GitHub: <redacted>

Repo URL: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>/tree/main

Commit ID: 

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely or locally)
iris = pd.read_csv("iris.csv")

# Output the first 15 rows of the data
print(iris.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(iris.info())

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The dataset contains measurements of iris flowers from three species—Setosa, Versicolor, and Virginica. Each row in the dataset represents one flower sample, with several measurements as features and the flower’s species as the label.

### Features
The features include:
1. **Sepal length** (in cm)
2. **Sepal width** (in cm)
3. **Petal length** (in cm)
4. **Petal width** (in cm)

### Labels
The label is the species of the iris flower, and each flower in the dataset belongs to one of three classes. This label is categorical, meaning it’s a name or category rather than a numerical value.

### Mapping of Labels to Classes
The dataset maps each species name to a specific class:
- **Setosa** (Class 0)
- **Versicolor** (Class 1)
- **Virginica** (Class 2)

# Part 2: Split the dataset into train and test

In [3]:
# Take the dataset and split it into our features (X) and label (y)
X = iris[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]
y = iris['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 42)

# Output the shape of the split data
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (135, 4)
Shape of X_test: (15, 4)
Shape of y_train: (135,)
Shape of y_test: (15,)


# Part 3: Logistic Regression

In [4]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression(max_iter = 200)
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
LR_sample_datapoint = [[4.7, 3.2, 1.3, 0.2]] # Class - Setosa
LR_predicted_probabilities = model.predict_proba(LR_sample_datapoint)
print("Predicted probabilities for each class:", LR_predicted_probabilities)

# iii. Report on the score for Logistic regression model, what does the score measure?
log_reg_score = model.score(X_test, y_test)
print("Logistic Regression Model Score:", log_reg_score)

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = model.coef_
intercepts = model.intercept_

print("Coefficients:", coefficients)
print("Intercepts:", intercepts)

Predicted probabilities for each class: [[9.83395192e-01 1.66047864e-02 2.17119200e-08]]
Logistic Regression Model Score: 1.0
Coefficients: [[-0.42830334  0.96729205 -2.44742526 -1.03780396]
 [ 0.5130202  -0.22029888 -0.21604909 -0.84453741]
 [-0.08471686 -0.74699317  2.66347435  1.88234137]]
Intercepts: [  9.54356486   1.89887279 -11.44243765]


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


### Explanation:
The score for the Logistic Regression model represents the model’s accuracy on the test set. Accuracy measures the proportion of correctly predicted labels out of the total number of predictions. A score close to 1 suggests a good fit, while a score close to 0 suggests a poor fit.

An accuracy score of 1.0 (or 100%) means the model correctly classified every sample in the test set.

### ii. Predicted Probabilities: 
   - The output [[9.83395192e-01 1.66047864e-02 2.17119200e-08]] indicates the predicted probabilities for each class in the order of the labels:
     - Class 0 **(Setosa)**: Approximately 98.34% probability
     - Class 1 **(Versicolor)**: Approximately 1.66% probability
     - Class 2 **(Virginica)**: Approximately 0.000002% probability
   - This means the model is highly confident that the sample belongs to the *Setosa* class.

### iii. Score:
The score function for Logistic Regression measures the percentage of correct predictions out of all predictions made on the test data. Given an output score of 1.0, the Logistic Regression model achieved 100% accuracy, meaning it correctly classified all test samples. This high accuracy indicates that the linear decision boundaries in logistic regression effectively captured the separable nature of the Iris dataset.

### iv. Equations
#### Setosa:
${y_1 = 9.54356486 - 0.42830334 \cdot \text{sepal length} + 0.96729205 \cdot \text{sepal width} - 2.44742526 \cdot \text{petal length} - 1.03780396 \cdot \text{petal width}}$

#### Versicolor:
${y_2 = 1.89887279 + 0.5130202 \cdot \text{sepal length} - 0.22029888 \cdot \text{sepal width} - 0.21604909 \cdot \text{petal length} - 0.84453741 \cdot \text{petal width}}$

#### Virginica:
${y_3 = -11.44243765 - 0.08471686 \cdot \text{sepal length} - 0.74699317 \cdot \text{sepal width} + 2.66347435 \cdot \text{petal length} + 1.88234137 \cdot \text{petal width}}$

# Part 4: Support Vector Machine

In [5]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc_model = SVC(probability = True)  # enables probability prediction
svc_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = [[4.7, 3.2, 1.3, 0.2]] # Class - Setosa
SVM_predicted_probabilities = svc_model.predict_proba(sample_datapoint)
print("Predicted probabilities for each class:", SVM_predicted_probabilities)

# iii. Report on the score for the SVM, what does the score measure?
score = svc_model.score(X_test, y_test)
print("Support Vector Classifier Score:", score)

Predicted probabilities for each class: [[0.97617484 0.01519442 0.00863074]]
Support Vector Classifier Score: 1.0


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


### ii. Predicted Probabilities: 
   - The output  [[0.97585978 0.01536693 0.00877329]] indicates the predicted probabilities for each class in the order of the labels:
     - Class 0 **(Setosa)**: Approximately 97.59% probability
     - Class 1 **(Versicolor)**: Approximately 1.54% probability
     - Class 2 **(Virginica)**: Approximately 0.87% probability
   - This means the model is highly confident that the sample belongs to the *Setosa* class.

### iii. Score:
The score here also calculates the accuracy (ratio of correctly classified instances over the total number of instances in the test set). An accuracy score of 1.0 here shows that the SVM model successfully created a hyperplane that separated all classes in the dataset without any errors on the test set.

# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(random_state = 42, max_iter = 1000)
mlp_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
NN_sample_datapoint = [[6.4, 2.7, 5.3, 1.9]]  # Class - Virginica

NN_predicted_probabilities = mlp_model.predict_proba(NN_sample_datapoint)
print("Predicted probabilities for each class:", NN_predicted_probabilities)

# iii. Report on the score for the Neural Network, what does the score measure?
mlp_score = mlp_model.score(X_test, y_test)
print("Neural Network Score:", mlp_score)

# iv: Experiment with different options for the neural network, report on your best configuration
best_score = 0
best_config = None

for hidden_layers in [(100,), (100, 50), (50, 50), (150,)]: # number of neurons in each hidden layer
    for activation in ['relu', 'tanh', 'logistic']: # different activation functions for the hidden layers 
        mlp_model = MLPClassifier(hidden_layer_sizes=hidden_layers, activation=activation, random_state=42, max_iter=1000)
        mlp_model.fit(X_train, y_train)
        score = mlp_model.score(X_test, y_test)
        
        if score > best_score:
            best_score = score
            best_config = (hidden_layers, activation)

print("Best Configuration:", best_config)
print("Best Score:", best_score)

/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


Predicted probabilities for each class: [[1.72341360e-06 1.91612909e-02 9.80836986e-01]]
Neural Network Score: 1.0
Best Configuration: ((100,), 'relu')
Best Score: 1.0


### ii. Predicted Probabilities: 
   - The output [[1.72341360e-06 1.91612909e-02 9.80836986e-01]] indicates the predicted probabilities for each class in the order of the labels:
     - Class 0 **(Setosa)**: Approximately 0.000172341360% probability
     - Class 1 **(Versicolor)**: Approximately 1.91% probability
     - Class 2 **(Virginica)**: Approximately 98.0836986% probability
   - This means the model is highly confident that the sample belongs to the *Virginica* class.

### iii. Score: 
The score here also measures the model’s accuracy. With a score of 1.0, this accuracy indicates that the neural network correctly classified all test samples.

### iv. Experimentation: 
   - Hidden Layer Sizes: (100,),
   - Activation Function: 'relu'
   
In this configuration, the neural network achieved the highest accuracy (1.0), demonstrating that a single hidden layer with 100 neurons and a ReLU activation function was sufficient for accurately classifying the dataset. The ReLU activation must have helped the network model complex, non-linear decision boundaries.

# Part 6: K-Nearest Neighbors

In [7]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn_model = KNeighborsClassifier(n_neighbors = 3)
knn_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
KNN_sample_datapoint = [[5.2,2.7,3.9,1.4,]]  # Class - Versicolor

KNN_predicted_probabilities = knn_model.predict_proba(KNN_sample_datapoint)
print("Predicted probabilities for each class:", KNN_predicted_probabilities)

# iii. Report on the score for kNN, what does the score measure?
knn_score = knn_model.score(X_test, y_test)
print("kNN Model Score:", knn_score)

Predicted probabilities for each class: [[0. 1. 0.]]
kNN Model Score: 1.0


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


### ii. Predicted Probabilities: 
   - The output [[0. 1. 0.]] indicates the predicted probabilities for each class in the order of the labels:
     - Class 0 **(Setosa)**: Approximately 0% probability
     - Class 1 **(Versicolor)**: Approximately 100% probability
     - Class 2 **(Virginica)**: Approximately 0% probability
   - This means the model is highly confident that the sample belongs to the *Versicolor* class.

### iii. Score:
The score here also calculates accuracy. An accuracy score of 1.0 indicates that the kNN classifier correctly identified the class of each test point by analyzing the closest neighbors’ classes, effectively leveraging the dataset’s separable structure.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

### Conclusions and Takeaways

1. **Logistic Regression**: This model provided predicted probabilities close to the correct class and achieved an accuracy score of 1.0 on the test set. Logistic Regression handled the dataset well, as the classes in the Iris dataset are linearly separable, which this model is well-suited for.

2. **Support Vector Machine (SVM)**: The SVM model also scored a perfect accuracy of 1.0, assigning probabilities close to 1 for the correct class. SVMs are effective for this dataset since they maximize the margin between classes, particularly in a relatively small, well-separated dataset like this one.

3. **Neural Network (MLP Classifier)**: The neural network model achieved high confidence in the predicted probabilities, though it differed from the other models by predicting a higher probability for a different class in one case. Neural networks can be highly accurate for complex datasets, though they may not be necessary for this dataset due to its simplicity.

4. **k-Nearest Neighbors (kNN)**: The kNN model achieved perfect accuracy with \(k=3\), correctly predicting the class of each instance. Given the small size and clear separability of the dataset, kNN performed well because the dataset naturally fits a distance-based approach where similar measurements lead to similar classifications.

### Best-Performing Models

Each model achieved a perfect accuracy of 1.0 on the test set, but kNN and SVM seemed to perform particularly well in predicting the correct class with high confidence.

### Insights

I was surprised that all models achieved a perfect score, which may be due to the small size of the dataset rather than a reflection of each model's overall performance on more complex datasets. Additionally, the confidence of KNN and SVM in their predictions was good to see, although it was more or less expected. This exercise demonstrated the versatility of different classification approaches when each model is capable of achieving high accuracy with $
well-tuned parameters.